flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1352,84 +1352,73 @@ public void triggerCheckpointBarrier(
+                (t) -> failExternally(new RuntimeException("Error while confirming checkpoint", t)),
+                "Ignoring checkpoint commit notification for non-running task {}.");
     }
 
     public void notifyCheckpointAborted(
             final long checkpointID, final long latestCompletedCheckpointId) {
-        final AbstractInvokable invokable = this.invokable;
+        notifyCheckpoint(
+                checkpointID,
+                (invokable) ->
+                        invokable.notifyCheckpointAbortAsync(
+                                checkpointID, latestCompletedCheckpointId),
+                "Notify checkpoint abort {} for {} ({}) was rejected by the mailbox",
+                "Error while aborting checkpoint {}.",
+                (t) -> failExternally(new RuntimeException("Error while aborting checkpoint", t)),
+                "Ignoring checkpoint aborted notification for non-running task {}.");
+    }
 
-        if (executionState == ExecutionState.RUNNING && invokable != null) {
-            try {
-                invokable.notifyCheckpointAbortAsync(checkpointID, latestCompletedCheckpointId);
-            } catch (RejectedExecutionException ex) {
-                // This may happen if the mailbox is closed. It means that the task is shutting
-                // down, so we just ignore it.
-                LOG.debug(
-                        "Notify checkpoint abort {} for {} ({}) was rejected by the mailbox",
-                        checkpointID,
-                        taskNameWithSubtask,
-                        executionId);
-            } catch (Throwable t) {
-                if (getExecutionState() == ExecutionState.RUNNING) {
-                    // fail task if checkpoint aborted notification failed.
-                    failExternally(new RuntimeException("Error while aborting checkpoint", t));
-                }
-            }
-        } else {
-            LOG.info(
-                    "Ignoring checkpoint aborted notification for non-running task {}.",
-                    taskNameWithSubtask);
-        }
+    public void notifyCheckpointSubsumed(long checkpointID) {
+        notifyCheckpoint(
+                checkpointID,
+                (invokable) -> invokable.notifyCheckpointSubsumedAsync(checkpointID),
+                "Notify checkpoint subsume {} for {} ({}) was rejected by the mailbox",
+                "Error while subsuming checkpoint {}.",
+                null,
+                "Ignoring checkpoint subsume notification for non-running task {}.");
     }
 
-    public void notifyCheckpointSubsumed(long checkpointId) {
+    private void notifyCheckpoint(
+            long checkpointId,
+            Consumer<AbstractInvokable> invokableNotifier,
+            String logFormatIfRejected,
+            String logFormatIfFailedToNotify,
+            @Nullable Consumer<Throwable> handleThrowableIfTaskRunning,
+            String logFormatIfNotRunning) {
```

Comment:
For me this method has the same problem as in my previous comment with generic Consumer's. However here I would propose to add an `enum NotifyCheckpointOperation` with three values `ABORT`, `COMPLETE`, `SUBSUME`, and just switch/case it to replace `Consumer<AbstractInvokable> invokableNotifier`.

On the other hand `handleThrowableIfTaskRunning` could be replaced by boolean flag `boolean shouldFailExternallyOnException`, so the signature of this method could be:
```
    private void notifyCheckpoint(
            long checkpointId,
            NotifyCheckpointOperation notifyCheckpointOperation,
            boolean shouldFailExternallyOnException)
```

Thirdly, the log/error formats I think could be handled also easier, by just using something like:
```
String.format("Notify checkpoint {} {} for {} ({}) was rejected by the mailbox", notifyCheckpointOperation.toString(), ...);
```
instead of
```
String.format("Notify checkpoint abort {} for {} ({}) was rejected by the mailbox",  ...);
```

Reply:
Thanks for the suggestion of introducing `enum` of `NotifyCheckpointOperation`, I think this could help reduce code duplication and also make the code more readable.